docs/proposals/circuit-breaker.md

```diff
@@ -0,0 +1,102 @@
+# Circuit Breaker
+
+## Context
+
+### Design pattern
+
+Circuit breaker is a popular design pattern in a distributed system world. It protects caller from long hanging in timeout if something unexpectedly goes wrong on the server side. Circuit breaker has 2 states - OPEN and CLOSED (sometimes also HALF OPEN). An initial state is CLOSED and all requests are proxied successfully to the server. If some threshold is reached (success rate or several consecutive errors) then circuit breaker switches to OPEN state. In OPEN state circuit breaker returns error immediately without an actual requesting of the server. Eventually, after a while, circuit breaker switches to CLOSED state again. 
+
+### Envoy support 
+
+Envoy allows to configure `cluster.CircuitBreakers` on each Cluster:
+```json
+{
+  "thresholds": []
+}
+```
+Where `thresholds` is an array of elements:
+```json
+{
+  "priority": "...",
+  "max_connections": "{...}",
+  "max_pending_requests": "{...}",
+  "max_requests": "{...}",
+  "max_retries": "{...}",
+  "retry_budget": "{...}",
+  "track_remaining": "...",
+  "max_connection_pools": "{...}"
+}
```

Comment:
This configuration is also important and can be a blocker when adopting Kuma. Especially `max_connections`